docs/RFCS/postgres_syntax.md

```diff
@@ -0,0 +1,98 @@
+- Feature Name: Postgres Syntax Compatibility
+- Status: draft
+- Start Date: 2017-04-16
+- Authors: Matt Jibson
+- RFC PR:
+- Cockroach Issue:
+
+# Summary
+
+This proposal is to define the specific goals of the two syntax types
+(traditional and modern) so we can correctly decide how to triage parsing
+differences between CockroachDB and Postgres, Spanner, or other databases.
+
+# Motivation
+
+The differences in syntax types (`traditional` and `modern`) in CockroachDB
+are not documented and have no defined goals or reasons for existing or
+differing. The unwritten reasons are that traditional is the old Postgres
+style and modern is the new Spanner style. However, we have already
+made some choices to differ from Postgres ([integer division returning
+decimals](https://github.com/cockroachdb/cockroach/pull/3308), for example)
+because we thought the Postgres syntax could be improved. This style worked
+fine for a while.
+
+With our recent goal of supporting various ORMs, these previous
+decisions become more complicated. We are advertising ourselves
+as an implementer of the Postgres protocol and syntax, and we
+mostly do. There is some syntax we do not yet support, and we
+rightly return errors for that. However we now have various ([integer
+division](https://github.com/cockroachdb/cockroach/pull/3308), [bitwise xor
+(~) operator](https://github.com/cockroachdb/cockroach/pull/14908), [shift
+argument modulo](https://github.com/cockroachdb/cockroach/issues/14874)) query
+expressions that are known to return different results that Postgres. In all
+of those cases we decided to differ because we believe the Postgres syntax
+to be flawed.
+
+Meanwhile, the modern syntax. It is undocumented, but it appears to
+differ in 3 ways from the traditional syntax: identifiers, strings, and
+comments. These all occur in the lexing phase of the parser. Our current
+implementation does not allow us to parse differently (for example, change
+the precedence of the `^` operator which is exponent in Postgres and xor
+in Spanner). Some syntax we had was inspired from Spanner, but because it
+differed too much from Postgres we had to remove or change it ([precedence
+of |, #, &](https://github.com/cockroachdb/cockroach/pull/14944), [^ is
+power instead of xor](https://github.com/cockroachdb/cockroach/pull/14882)).
+
+Overall we are moving away from both the Postgres and Spanner syntaxes and
+making our own. In our support of ORMs, we will have to document an ongoing
+list of differences between us and Postgres, and we will never be able to
+well support the Spanner syntax.
```

Comment:
Why do you se "we will  never..."

If the concern is about *syntax*, then the solution is to implement two *grammars* side-by-side. This is rather feasible implementation-wise. If the concern is about *semantics* then the solution is to implement separate *dialects* side-by-side. This is more complicated but there is precedent in other technology.